File: lib/ch4/4.4.js
// Given a binary search tree, design an algorithm which creates
// a linked list of all the nodes at each depth (i e , if you have a
// tree with depth D, you’ll have D linked lists)
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class BST {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

const createBuckets = (root, buckets = {}, depth = 1)  => {
  if (root === null) return buckets;

  if (!buckets[depth]) buckets[depth] = [];
  buckets[depth].push(root.value);
  createBuckets(root.left, buckets, depth + 1);
  createBuckets(root.right, buckets, depth + 1);

  return buckets;
};

const createLists = buckets => {
  const lists = [];

  Object.keys(buckets).forEach(k => {
    const dummy = new Node(null);
    let node = dummy;
    const bucket = buckets[k];
    bucket.forEach(val => {
      node.next = new Node(val);
      node = node.next;
    });
    lists.push(dummy.next);
  });

  return lists;
};

File: java/ch4/4.4.java
class TreeNode {
    int data;
    TreeNode left;
    TreeNode right;
    public TreeNode(int data) {
        data = data;
        left = right = null;
    }
}

class LinkedList {
    int data;
    LinkedList next;
    public LinkedList(int data) {
        data = data;
        next = null;
    }
}

void createLevelLinkedList(TreeNode root,
                        ArrayList<LinkedList<TreeNode>> lists, int level) {
    if (root == null) return;
    LinkedList<TreeNode> list = null;
    if (lists.size() == level) {
        list = new LinkedList<TreeNode>();
        lists.add(list);
    } else {
        list = lists.get(level);
    }
    list.add(root);
    createLevelLinkedList(root.left, lists, ++level);
    createLevelLinkedList(root.right, lists, ++level);
}

ArrayList<LinkedList<TreeNode>> createLevelLinkedLists(TreeNode root) {
    ArrayList<LinkedList<TreeNode>> lists = new ArrayList<LinkedList<TreeNode>>();
    createLevelLinkedList(root, lists, 0);
    return lists;
}
